Archive/Imo/Imo2025Q3.lean

```diff
@@ -0,0 +1,224 @@
+/-
+Copyright (c) 2025 Yi Yuan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yi Yuan
+-/
+import Mathlib.NumberTheory.LSeries.PrimesInAP
+import Mathlib.NumberTheory.Multiplicity
+import Mathlib.Tactic.Simproc.Factors
+
+/-!
+# IMO 2025 Q3
+
+Let $\mathbb{N}+$ denote the set of positive integers. A function $f: \mathbb{N}+ \rightarrow
+\mathbb{N}+$ is said to be bonza if $f(a)$ divides $b^a-f(b)^{f(a)}$ for all positive integers $a$
+and $b$. Determine the smallest real constant $c$ such that $f(n) \leq c n$ for all bonza functions
+$f$ and all positive integers $n$.
+
+## Solution
+
+We follow solution from https://web.evanchen.cc/exams/IMO-2025-notes.pdf.
+
+We first plug in `a = b = n` to get the basic constraint `∀ n, f n ∣ n ^ n`. Next, one shows that
+unless `f` is the identity, every odd prime must satisfy `f(p)=1`. From here, any odd prime divisor
+of `f(n)` is ruled out by taking `a = n, b = p`, so `f(n)` is always a power of `2`.
+Finally, evaluating `a = n, b = 3` gives `f n ∣ 3 ^ n - 1`, and according to the LTE lemma, we have
+`padicValNat 2 (3 ^ n - 1) = padicValNat 2 n + 2` for `Even n`. Therefore,
+`f(n) ≤ 2 ^ (padicValNat 2 n + 2) ≤ 4 * n`, so `c=4` works.
+
+A matching construction is `f(n)=1` for `Odd n`, `f(4)=16`, and `f(n)=2` for other `Even n`, which
+attains the bound, showing the optimal answer is `c=4`.
+-/
+
+open Nat Int
+
+namespace Imo2025Q3
+
+/-- Define bonza functions -/
+def bonza : (ℕ → ℕ) → Prop := fun (f : ℕ → ℕ) ↦
+  (∀ a b : ℕ, 0 < a → 0 < b → (f a : ℤ) ∣ (b : ℤ) ^ a - (f b : ℤ) ^ (f a)) ∧ ∀ n, 0 < n → 0 < f n
+
+namespace bonza
+
+variable {f : ℕ → ℕ}
+
+/-- For each bonza function $f$, we have $f n | n ^ n$ -/
```

Comment:
Here and elsewhere, I think we generally prefer use of Markdown-style backticks plus Unicode rather than `$ $` around mathematical expressions. And that's definitely the case when the expression is in fact given in Lean notation rather than mathematical LaTeX, as here (using `f n` rather than $f(n)$).